Issue/140

tschob opened the pull request:
Implements multiple instances support. Solves #140 

There is now a third example project which shows two instances in parallel.

kevindelord committed:
Storyboard example now full screen. Update podfile

kevindelord committed:
Update gitignore

kevindelord committed:
Add TODOs and replace the shared instance in the FolioReaderKit class

kevindelord committed:
Refactor the readerCenter and global book entity

kevindelord committed:
ReaderContainer: improve usage

kevindelord committed:
Refactor reader config

kevindelord committed:
Improve isDirection functions to use current readerConfig

kevindelord committed:
PageScrubbler: improve init logic and remove few optional unwrap

kevindelord committed:
Refactor highlightlist, scrollscrubber and chapter list

kevindelord committed:
Refactor FolioReaderFontsMenu

kevindelord committed:
Refactor FolioReaderPlayerMenu

kevindelord committed:
Refactor FolioReaderQuoteShare

kevindelord committed:
Refactor FolioReaderAudioPlayer

kevindelord committed:
Refactor FolioReaderWebView.swift

kevindelord committed:
Highlight: remove explicit try!

kevindelord committed:
Database method: integrate dynamic reader configuration

kevindelord committed:
Refactor HighlightStyle

kevindelord committed:
Remove few todos and minor code improvment

kevindelord committed:
Merge pull request #5 from smartmobilefactory/highlight_refactor

Highlight refactor

kevindelord committed:
remove optional for book and epubPath

kevindelord committed:
Remove unnecessary optional for FRBook

kevindelord committed:
Update todos

kevindelord committed:
PR: add doc and TODOs

kevindelord committed:
ignoreSystemTint: remove shared readerContainer and many optional unwrapping

kevindelord committed:
Reafactor FolioReaderChapterListCell

kevindelord committed:
Remove todo

kevindelord committed:
Merge pull request #6 from smartmobilefactory/refactor_reader_config

Refactor reader config

kevindelord committed:
Carthage; Update gitignore

kevindelord committed:
Merge branch 'refactor_reader_config' into remove_global_variables

kevindelord committed:
FolioReader: convert convenience init to internal

kevindelord committed:
Highlight: create extension with all static deprecated functions inside

kevindelord committed:
Remove kBookId global variable

kevindelord committed:
Remove calls to FolioReader.shared.ReaderCenter

kevindelord committed:
Example app: integrate new API and refactor code

kevindelord committed:
FolioReader: remove all use of FolioReader.shared attribute.

kevindelord committed:
Remove todo

kevindelord committed:
Add TODO

kevindelord committed:
PR: refactor guard

kevindelord committed:
Merge branch 'remove_global_variables' into remove_shared_instances

kevindelord committed:
Refactor: isScrolling

kevindelord committed:
Refactor: pageScrollDirection

tschob committed:
Merge pull request #7 from smartmobilefactory/remove_global_variables

Remove global variables

Commit:
Merge branch 'remove_shared_instances' of github.com:smartmobilefactory/FolioReaderKit into development

kevindelord committed:
Refactor: currentPageNumber

kevindelord committed:
Refactor: reuseIdentifier

kevindelord committed:
Refactor: pageheight and pageWidth

kevindelord committed:
Refactor: remove possible division by 0

kevindelord committed:
Refactor: remove some optional unwrappings

kevindelord committed:
Refactor: function with bad structure

kevindelord committed:
Refactor: duplicated code

kevindelord committed:
Merge branch 'remove_shared_instances' of https://github.com/smartmobilefactory/FolioReaderKit into remove_shared_instances

kevindelord committed:
Merge branch 'remove_shared_instances' into readerCenter_global_variables

kevindelord committed:
Remove optional unwrapping

kevindelord committed:
Merge pull request #9 from smartmobilefactory/readerCenter_global_variables

Reader center global variables

kevindelord committed:
Fix and better integrate the storyboard support.

kevindelord committed:
Add todo check

kevindelord committed:
Update API logic to return FolioReaderContainer

kevindelord committed:
Wrap all store values in dicionaries

kevindelord committed:
Add documentation

kevindelord committed:
Fix typo in delegate function name

kevindelord committed:
FolioReader: mark functions as deprecated

kevindelord committed:
Highlight: mark functions as deprecated

kevindelord committed:
Remove TODO_SMF_DOCs and create new extensions file

kevindelord committed:
Report more changes on the migration.md file

kevindelord committed:
Improve question comment text

kevindelord committed:
Comments: fix typo

kevindelord committed:
Delegate: deprecate one function

kevindelord committed:
Add more migration documentation

kevindelord committed:
update documentation and readme

kevindelord committed:
Add one minor 0 check

tschob committed:
Merge pull request #8 from smartmobilefactory/remove_shared_instances

Remove shared instances

Commit:
Merge branch 'refactor_userdefault' of github.com:smartmobilefactory/FolioReaderKit into development

tschob committed:
Merge pull request #11 from smartmobilefactory/refactor_userdefault

Refactor userdefault

kevindelord committed:
Migrationm: add description

kevindelord committed:
Update TODO_SMF_QUESTION

kevindelord committed:
Update migration.md

Commit:
Remove unwanted space

tschob committed:
Merge pull request #12 from smartmobilefactory/documentation_and_deprecations

Documentation and deprecations

Commit:
Remove unneeded SMF content

Commit:
Remove Xcode 8.1 logging workaround from the schemes

Commit:
Replace tabs with spaces for the indentation

tschob committed:
Merge pull request #14 from smartmobilefactory/documentation_and_deprecations

Replace tabs with spaces for the indentation

Commit:
Remove outcommented code

Commit:
Merge branch 'development' of github.com:smartmobilefactory/FolioReaderKit into development

Commit:
Fix indentation

Commit:
Remove out commented code

Commit:
Replace tabs with spaces for the indentation

Commit:
Replace tabs with spaces for the indentation

Commit:
Replace tabs with spaces for the indentation

tschob committed:
Merge pull request #13 from smartmobilefactory/development

Development

Commit:
Merge branch 'master' of github.com:FolioReader/FolioReaderKit into folio/issue/140

Conflicts:
	Example/Example/ViewController.swift
	Source/EPUBCore/FREpubParser.swift
	Source/FolioReaderCenter.swift
	Source/FolioReaderContainer.swift
	Source/FolioReaderFontsMenu.swift
	Source/FolioReaderKit.swift
	Source/FolioReaderPage.swift
	Source/FolioReaderWebView.swift
	Source/Models/Highlight+Helper.swift

tschob committed:
Merge pull request #15 from smartmobilefactory/merge

Merge

Commit:
Remove unused enum case

Commit:
Fix "getCurrentChapterName()"

Commit:
Remove SMF TODOs which are not questions

Commit:
Refactor reader config identifier logic to use custom identifiers instead of the book title and support user default settings from former app installations

Commit:
Add multiple instances with Storyboard example

Commit:
Settings: Register default values only if a key wasn't registered before and add a prefix to the identifier

Commit:
Fix night mode appearance on first book opening

Commit:
Add multiple instances code base example project

kevindelord committed:
Add missing shared instance configuration. Remove optional unwrapping and related crash DBGIOS-934

Commit:
Fix indentation with tabs to use spaces

tschob committed:
Merge pull request #16 from smartmobilefactory/fix_legacy

Add missing shared instance configuration. Remove optional unwrapping…

Commit:
Fix wrong page number when using horizontal with vertical content scrolling

Commit:
Merge branch 'master' of github.com:smartmobilefactory/FolioReaderKit into multipleInstancesMerge

Commit:
Really fix wrong page number when using horizontal with vertical content scrolling

Commit:
Fix sometimes wrong page offset for new pages when using vertical with horizontal content scroll direction

Commit:
Merge branch 'master' of github.com:smartmobilefactory/FolioReaderKit into multipleInstancesMerge

Commit:
Improved TODO and FIXME naming

Commit:
Fix indentation

tschob committed:
Merge pull request #17 from smartmobilefactory/multipleInstancesMerge

Multiple instances merge

arctouch-rodrigoalmeida commented on Source/EPUBCore/FRMediaType.swift:
```diff
@@ -9,21 +9,21 @@
-    
+    var name                : String
+    var defaultExtension    : String!
+    var extensions          : [String]!
```
Why are you using `Forced Unwrapping` here? All these properties are initialized in the `init`

kevindelord replied:
I am/We are not using `Forced Unwrapping` there. Here at the SmartMobileFactory this this strictly forbidden. Those unwrappings have been implemented before we started any change in the library.

On this pull request, while implementing the new feature I personally removed many `Forced Unwrapping` in favour of `Optional` ones. But the cost of such migration is pretty heavy as it takes a lot of times and could also create side-effect bugs.

kevindelord replied:
FYI: In most cases you can see those attribute declarations in the diff due to a re-indentation or adding white spaces (which will be removed as requested).

arctouch-rodrigoalmeida commented on Source/EPUBCore/FRMetadata.swift:
```diff
@@ -9,13 +9,13 @@
-    
+    var name        : String!
+    var role        : String!
+    var fileAs      : String!
```
Also here

arctouch-rodrigoalmeida commented on Source/EPUBCore/FRMetadata.swift:
```diff
@@ -39,34 +39,34 @@ struct Identifier {
+    var date        : String!
+    var event       : String!
+
     init(date: String, event: String!) {
```
Using `String!` isn't good, maybe change to `event: String?` will be better

arctouch-rodrigoalmeida commented on Source/EPUBCore/FRResource.swift:
```diff
@@ -9,12 +9,12 @@
+    var mediaType       : MediaType!
+    var mediaOverlay    : String?
 
     func basePath() -> String! {
```
Also here

arctouch-rodrigoalmeida commented on Source/EPUBCore/FRSmilElement.swift:
```diff
@@ -13,9 +13,9 @@ import UIKit
-    var attributes: [String: String]!
-    var children: [FRSmilElement]
+    var name                : String // the name of the tag: <seq>, <par>, <text>, <audio>
+    var attributes          : [String: String]!
```
Here again, you don't need to use `!`

arctouch-rodrigoalmeida commented on Source/EPUBCore/FRSpine.swift:
```diff
@@ -9,18 +9,19 @@
-    var resource: FRResource!
-    
+    var linear          : Bool!
+    var resource        : FRResource!
```
Also here

arctouch-rodrigoalmeida commented on Source/EPUBCore/FRTocReference.swift:
```diff
@@ -9,15 +9,15 @@
-    var children: [FRTocReference]!
-    
+    var resource        : FRResource?
+    var title           : String!
```
Here too

arctouch-rodrigoalmeida commented on Source/FolioReaderChapterList.swift:
```diff
@@ -53,38 +69,44 @@ class FolioReaderChapterList: UITableViewController {
 
-                cell.indexLabel.text = cell.indexLabel.text! + (duration != nil ? " - "+durationFormatted! : "");
+                if let durationFormatted = (duration != nil ? duration : "")?.clockTimeToMinutesString() {
+                    let text = (cell.indexLabel?.text ?? "")
```
You don't need to use `parentheses`

arctouch-rodrigoalmeida commented on Source/FolioReaderChapterListCell.swift:
```diff
@@ -9,34 +9,50 @@
-    var indexLabel = UILabel()
-    
+
+    var indexLabel 			: UILabel?
```
This isn't a good practice (a lot of space)

https://github.com/raywenderlich/swift-style-guide

hebertialmeida replied:
I think this spaces on var type and naming are unnecessary.

arctouch-rodrigoalmeida commented on Source/FolioReaderContainer.swift:
```diff
@@ -9,59 +9,76 @@
+        self.folioReader.readerContainer = self
+
+        // Initialize the default reader options.
+        if (self.epubPath != "") {
```
parenthesis again

arctouch-rodrigoalmeida commented on Source/FolioReaderContainer.swift:
```diff
@@ -72,131 +89,135 @@ open class FolioReaderContainer: UIViewController {
-        if readerConfig.canChangeScrollDirection {
-            var scrollDirection = FolioReaderScrollDirection(rawValue: FolioReader.currentScrollDirection) ?? .vertical
+        if (self.readerConfig.canChangeScrollDirection == true) {
+            var scrollDirection = (FolioReaderScrollDirection(rawValue: self.folioReader.currentScrollDirection) ?? .vertical)
```
Also here

arctouch-rodrigoalmeida commented on Source/FolioReaderContainer.swift:
```diff
@@ -72,131 +89,135 @@ open class FolioReaderContainer: UIViewController {
-            if (scrollDirection == .defaultVertical && readerConfig.scrollDirection != .defaultVertical) {
-                scrollDirection = readerConfig.scrollDirection
+            if (scrollDirection == .defaultVertical && self.readerConfig.scrollDirection != .defaultVertical) {
+                scrollDirection = self.readerConfig.scrollDirection
```
https://github.com/raywenderlich/swift-style-guide

kevindelord replied:
I am not sure what you mean there. I know that style guide, and I also know that we do not follow it.
We actually branched off from that one and wrote our own.
You can find it there: https://github.com/kevindelord/swift-style-guide

As we would like to merge this code change into @hebertialmeida's repo, we will of course follow/do what you/him want(s) us to modify.

arctouch-rodrigoalmeida commented on Source/FolioReaderUserDefaults.swift:
```diff
@@ -0,0 +1,110 @@
+        if let _identifier = identifier {
+            self.identifier = "folioreader.userdefaults.identifier.\(_identifier)"
+        } else {
+            self.identifier = nil
```
You don't need this `else`

arctouch-rodrigoalmeida commented on Source/Models/Highlight.swift:
```diff
@@ -11,17 +11,17 @@ import RealmSwift
+    open dynamic var contentPost    : String!
+    open dynamic var contentPre     : String!
+    open dynamic var date           : Foundation.Date!
+    open dynamic var highlightId    : String!
```
You can use `optional` or initiate all these properties

arctouch-rodrigoalmeida commented on Source/QuoteImage.swift:
```diff
@@ -12,10 +12,10 @@ import UIKit
+    public var image            : UIImage!
+    public var alpha            : CGFloat!
+    public var textColor        : UIColor!
+    public var backgroundColor  : UIColor!
```
These properties will be `nil` sometime?

arctouch-rodrigoalmeida commented on Source/ScrollScrubber.swift:
```diff
@@ -28,100 +28,109 @@ func > <T : Comparable>(lhs: T?, rhs: T?) -> Bool {
-	}
-
-    init(frame:CGRect) {
+    weak var delegate                   : FolioReaderCenter!
```
You need to use `optional`

kevindelord replied:
As explained in the other thread, this was done beforehand not by us. 
Removing all `Forced Unwrapping` is not really in the scope of that pull request.

hebertialmeida commented on Source/FolioReaderContainer.swift:
```diff
@@ -9,59 +9,76 @@
+    var shouldHideStatusBar         = true
+    var shouldRemoveEpub            = true
+    var epubPath                    : String
+    var book                        : FRBook
```
The same applies here, lets keep the previous style:
```swift
var name: Type
var aLongName: Type
```
not objective-c style:
```swift
var name:      Type
var aLongName: Type
```

kevindelord replied:
Even though that is not our way at all, I removed every single unnecessary white spaces on the variable declarations in the code. The current comments will be "done" once we merge the related branch into the `master` one.

hebertialmeida commented on Source/FolioReaderCenter.swift:
```diff
@@ -9,120 +9,124 @@
     required public init?(coder aDecoder: NSCoder) {
-        super.init(coder: aDecoder)
-        initialization()
+        fatalError("This class doesn't support NSCoding.")
```
Why remove it?

kevindelord replied:
Because the `FolioReaderCenter` class must be initialised with non-nil `readerContainer: FolioReaderContainer` variable.

```swift
open var readerContainer: FolioReaderContainer

init(withContainer readerContainer: FolioReaderContainer) {
        self.readerContainer = readerContainer
        super.init(nibName: nil, bundle: Bundle.frameworkBundle())

        self.initialization()
}
```

Without a `readerContainer` the `FolioReaderCenter` class cannot be / should be instantiated.

hebertialmeida commented on Source/FolioReaderCenter.swift:
```diff
@@ -9,120 +9,124 @@
+
         // Loading indicator
-        let style: UIActivityIndicatorViewStyle = isNight(.white, .gray)
+        let style: UIActivityIndicatorViewStyle = self.readerContainer.folioReader.isNight(UIActivityIndicatorViewStyle.white, UIActivityIndicatorViewStyle.gray)
```
Lets use the . syntax for enum, it is not necessary the full name like `UIActivityIndicatorViewStyle.gray`, just `.gray`.

That is why we have types :)

hebertialmeida commented on Source/FolioReaderPage.swift:
```diff
@@ -75,200 +97,208 @@ open class FolioReaderPage: UICollectionViewCell, UIWebViewDelegate, UIGestureRe
+
+        let direction: ScrollDirection = self.folioReader.needsRTLChange ? .positive(withConfiguration: self.readerConfig) : .negative(withConfiguration: self.readerConfig)
+
+        if (self.folioReader.readerCenter?.pageScrollDirection == direction && self.readerConfig.scrollDirection != .horizontalWithVerticalContent) {
```
There is a bug here, when you go to table of contents and select a new chapter it should not scroll down as it doing now

kevindelord replied:
I can reproduce it, I will look into it :)

kevindelord replied:
Aaaaand that one is also fixed !

hebertialmeida commented on Source/FolioReaderCenter.swift:
```diff
@@ -335,259 +357,267 @@ open class FolioReaderCenter: UIViewController, UICollectionViewDelegate, UIColl
+            return
+        }
+
+        let shouldHide = (!self.navigationController!.isNavigationBarHidden)
```
Why parenthesis?

hebertialmeida commented on Source/FolioReaderCenter.swift:
```diff
@@ -335,259 +357,267 @@ open class FolioReaderCenter: UIViewController, UICollectionViewDelegate, UIColl
     }
-    
+
     func toggleBars() {
```
There is a bug on bar toggle, on the Sherlock Holmes book, tap to show is not showing the bar, it toggle fast but keeps hidden.

kevindelord replied:
I can reproduce it, I will look into it :)

kevindelord replied:
That one is fixed :)

hebertialmeida reviewed (changes requested):
Huuuge one...

Thanks @kevindelord and @tschob for this, in general looks pretty good, my only concerns are the code style, @arctouch-rodrigoalmeida made some comments and I also did. Testing that I found two bugs that are happening to me, the tabBar toggle and the scroll to bottom when change chapter from Table of Contents.

The other thing that I don't know is on MultipleInstance-Example the code one is not multiple instance, only the storyboard.

kevindelord committed:
PR: Remove all extra whitepace on variable declarations DBGIOS-1004

kevindelord commented on Example/MultipleInstances-Example/CodeExample/CodeExampleViewController.swift:
```diff
@@ -0,0 +1,135 @@
+class CodeExampleViewController: UIViewController {
+
+    @IBOutlet var bookOne   : UIButton?
+    @IBOutlet var bookTwo   : UIButton?
```
As they are declared as `@IBOutlet` the connection/link in the storyboard to the actual variable might not be set which would lead to a crash.

Plus I receive xcode building errors if I even try to do so.
Errors:
```
- @IBOutlet property has non-optional type 'UIButton'
- Class 'ViewController' has no initializers
```

kevindelord commented on Example/MultipleInstances-Example/CodeExample/CodeExampleViewController.swift:
```diff
@@ -0,0 +1,135 @@
+
+extension CodeExampleViewController {
+
+    @IBAction func didOpen(_ sender: AnyObject) {
```
I can't do that as we use the `tag` attribute of the parameter `sender`. The type `Any` does not have such attribute.
Building Error:
```
Value of type 'Any' has no member 'tag'
```

kevindelord committed:
PR: Remove parenthesis and some minor demands DBGIOS-1004

Commit:
Fix indentation with tabs to use spaces

tschob committed:
Merge pull request #18 from smartmobilefactory/fix_PR_comments

Fix PR comments DBGIOS-1004

kevindelord committed:
Go to table of contents and select a new chapter, it should not scroll down DBGIOS-1004

kevindelord committed:
Bug fix: Bar toggle: tap to show is now showing the bar DBGIOS-1004

kevindelord committed:
Merge branch 'fix_PR_comments' of https://github.com/smartmobilefactory/FolioReaderKit into fix_PR_comments

kevindelord committed:
PR: Remove extra brackets

kevindelord committed:
Bug fix: jumping to chapter DBGIOS-1004

kevindelord committed:
Remove commented code DBGIOS-1004

kevindelord committed:
Add epub identifier to avoid problems between example books DBGIOS-1004

kevindelord committed:
PR: minor indentation change

kevindelord commented:
Hi @hebertialmeida !

So an internal PR will soon be reviewed and merged. It will fix the remaining issues and bugs.
Could you please try out the app again and found new introduce bugs?

Best!

kevindelord commented:
Hi again @hebertialmeida,

I'm opening another thread for that small question. 
When using `horizontalWithVerticalContent`, once the user jumps to another chapter (from the chapter list) he is not redirected to the top of the page (with the chapter title) but somewhere else.
This behaviour differs from the `vertical` and `horizontal` scroll directions.

I am not sure if this is a bug or feature.
I've also made sure that I have not introduced that problem (if it is one).

Best

Commit:
Merge branch 'fix_PR_comments' of github.com:smartmobilefactory/FolioReaderKit into fork/master

Commit:
Fix indentation with tabs to use spaces

Commit:
Update to CocoaPods version 1.2.1

tschob committed:
Merge pull request #19 from smartmobilefactory/fix_PR_comments

Fix PR comments DBGIOS-1004

hebertialmeida commented:
So, the `horizontalWithVerticalContent` should behave like that.

hebertialmeida committed:
Fixed bar animation and not showing after couple page loads

The bar was not showing because now with `setup` function it add
multiple times the recognizer to the view, now I remove everything
before adding new one

hebertialmeida committed:
Light refactor to have weak references of readerContainer

hebertialmeida commented:
Hey, @tschob and @kevindelord I take a general look at it, improved the animation of tabBar and fixed the bug of not showing it.

The problem was that the gestured is added multiple times to the webView, so when you tap it to toggle the bar the number of times the gestures was added...

Overall it LGTM, so I think we can merge that, take a look at my changes and let me know., then I will merge it.

hebertialmeida committed:
Fixed tests

kevindelord commented:
@hebertialmeida I don't fully understand why you changed the readerContainer to an optional:
```swift
open weak var readerContainer: FolioReaderContainer?
```
But otherwise it looks good to me. You can merge :)

kevindelord commented on Source/FolioReaderCenter.swift:
```diff
@@ -70,12 +71,19 @@ open class FolioReaderCenter: UIViewController, UICollectionViewDelegate, UIColl
-        return self.readerContainer.readerConfig
+    fileprivate var folioReader: FolioReader {
+        guard let readerContainer = readerContainer else { return FolioReader() }
+        return self.readerContainer!.folioReader
```
like if you had kept the non-optional value this would not have been needed.
Plus you should use the value from the `guard` instead of the `self.readerContainer!.folioReader`

hebertialmeida replied:
Yeah, missed that...

hebertialmeida committed:
Removed forcing object

hebertialmeida committed:
Use weak reference

hebertialmeida committed:
Removed the need for retain ReaderContainer

kevindelord commented on Source/FolioReaderKit.swift:
```diff
@@ -428,22 +442,6 @@ extension FolioReader {
-
-    /// Called when the application will resign active
-    open class func applicationWillResignActive() {
-        FolioReader.shared.saveReaderState()
```
Maybe we should keep that for the legacy? I mean when people update the library they could/should get a simple warning. 
The function could be empty though, but at least it would still exist :)

hebertialmeida replied:
As per discussed by Slack, there is no need for this.

kevindelord commented on Source/FolioReaderKit.swift:
```diff
@@ -428,22 +442,6 @@ extension FolioReader {
-
-    /// Called when the application will terminate
-    open class func applicationWillTerminate() {
-        FolioReader.shared.saveReaderState()
```
Same remark here :)

hebertialmeida replied:
As per discussed by Slack, there is no need for this.

kevindelord commented on Source/FolioReaderKit.swift:
```diff
@@ -127,9 +130,22 @@ open class FolioReader: NSObject {
+    fileprivate func addObservers() {
+        removeObservers()
+        NotificationCenter.default.addObserver(self, selector: #selector(saveReaderState), name: .UIApplicationWillResignActive, object: nil)
+        NotificationCenter.default.addObserver(self, selector: #selector(saveReaderState), name: .UIApplicationWillTerminate, object: nil)
```
Are you sure this works fine with the multiple instances mode? Like the correct `epub` gets saved?

hebertialmeida replied:
It should because each observer is added to `self` and it is a different object in memory.

hebertialmeida commented:
@kevindelord @tschob Thank you guys, this is a huge improvement!!

hebertialmeida committed:
Updated migration guide

hebertialmeida committed:
Make some methods public

hebertialmeida committed:
Updated dependencies
